src/components/plot/plot-control-grid.tsx

```diff
@@ -0,0 +1,169 @@
+import '@isrd-isi-edu/deriva-webapps/src/assets/scss/_heatmap.scss';
+
+// hooks
+import { useEffect, useState } from 'react';
+import '/node_modules/react-resizable/css/styles.css';
+import '/node_modules/react-grid-layout/css/styles.css';
+// services
+import ChaiseSpinner from '@isrd-isi-edu/chaise/src/components/spinner';
+import { DataConfig, defaultGridProps } from '@isrd-isi-edu/deriva-webapps/src/models/plot';
+import { convertKeysSnakeToCamel, validateGridProps } from '@isrd-isi-edu/deriva-webapps/src/utils/string';
+import { LayoutConfig } from '@isrd-isi-edu/deriva-webapps/src/models/plot';
+import { setControlData, useControl } from '@isrd-isi-edu/deriva-webapps/src/hooks/control';
+import { Layouts, Responsive, WidthProvider } from 'react-grid-layout';
+import ChartWithEffect from '@isrd-isi-edu/deriva-webapps/src/components/plot/chart-with-effect';
+import UserControl from '@isrd-isi-edu/deriva-webapps/src/components/plot/user-control';
+
+
+
+export type PlotControlGridProps = {
+  config: DataConfig,
+};
+const ResponsiveGridLayout = WidthProvider(Responsive);
+/**
+ * 
+ * @param config config object for the given plot
+ * @returns 
+ */
+const PlotControlGrid = ({
+  config,
+}: PlotControlGridProps): JSX.Element => {
+
+  const [layout, setLayout] = useState<Layouts>({});
+  const [gridProps, setGridProps] = useState({});
+  const [dataOptions, setDataOptions] = useState<any>(null);
+  const [userControlsExists, setUserControlExists] = useState<boolean>(false);
+  const [userControlsReady, setUserControlReady] = useState<boolean>(false);
+  const [initialParams, setInitialParams] = useState<any>({
+    $url_parameters: {
+      Gene: {
+        data: {
+          NCBI_GeneID: 1, // TODO: deal with default value
+        },
+      },
+      Study: [],
+    },
+    $control_values: {},
+    noData: false
+  });
+
+  useEffect(() => {
+    if (config) {
+      //If layout is configured use the given layout
+      if (config?.layout && Object.values(config?.layout).length > 0) {
+        const mappedLayoutValues = Object.values(config?.layout)?.map((resLayout: any) => (
+          resLayout.map((item: LayoutConfig) => convertKeysSnakeToCamel(({
+            //i defines the item on which the given layout will be applied
+            i: item?.source_uid,
+            ...item,
+          })))
+        ));
+        setLayout(Object.fromEntries(Object.entries(config.layout).map(([key]: any, index) => [key, mappedLayoutValues[index]])));
+      }
+      //Otherwise set the default layout to display controls and plots 
+      else {
+        const defaultPlotUid = config.plots.map((plot) => {
```

Comment:
If there is no `uid` defined in a `plots[x]` configuration, we should use a default of `plot_type` and the index in `plots`. So something like `bar_0` or `violin_0` for some of our existing configurations.